File: IOS/Packages/LiquorData/Sources/LiquorData/Repositories/CollectionExport.swift
import Foundation
import GRDB
import LiquorEngine

/// Exports the whole collection as CSV.
///
/// **Free, always, and prominent.** People in this category have been burned —
/// apps losing collections, accounts vanishing, years of entry gone — and the
/// advice they give each other is to be wary of anything that will not let you
/// export. It costs almost nothing to build and it is the strongest trust
/// signal available.
///
/// It also has to be genuinely complete. An export missing the barrel fields
/// would be worse than none: it would look like a backup and not be one.
public struct CollectionExport: Sendable {
    private let db: AppDatabase

    public init(_ db: AppDatabase) { self.db = db }

    /// Column order is deliberate: identity, then status, then the barrel
    /// detail, then money, then notes. It reads left to right the way somebody
    /// would describe a bottle out loud.
    public static let header = [
        "name", "distillery", "brand", "expression",
        "class", "production", "status",
        "sample", "sample_from", "sample_source", "infinity",
        "size_ml", "abv", "proof", "chill_filtered",
        "batch", "barrel", "store_pick", "pick_group", "pick_store",
        "warehouse", "rick", "floor", "recipe_code",
        "age_months", "entry_proof", "char_level", "finish",
        "bottle_number", "bottles_in_batch", "topper_letter", "wax_color", "drip_percent", "stave_recipe", "dsp",
        "distilled_year", "bottled_year", "vintage_year",
        "purchased", "opened", "killed", "last_poured", "last_verified",
        "price", "bought_at", "storage_location", "shelf_number",
        "pours_remaining", "pours_total", "ml_remaining", "cost_per_pour",
        "rating", "would_rebuy", "finish_seconds", "liked", "disliked",
    ]

    public func csv(
        resolveName: (Bottle) -> String,
        resolveIdentity: (String) -> ProductIdentity?
    ) throws -> String {
        let bottles = BottleRepository(db)
        let tastings = TastingRepository(db)

        // Killed bottles are INCLUDED. They are archived, not deleted, and a
        // backup that silently drops your history is not a backup.
        let summaries = try bottles.summaries(includeFinished: true)

        let rows: [[String]] = try summaries.map { summary in
            let bottle = summary.bottle
            let identity = bottle.catalogProductId.flatMap(resolveIdentity)
            let latest = try tastings.history(bottleId: bottle.id).first

            return [
                resolveName(bottle),
                CSVWriter.text(identity?.distillery),
                CSVWriter.text(identity?.brand),
                CSVWriter.text(identity?.expression),
                CSVWriter.text(identity?.classType.label),
                CSVWriter.text(identity?.productionType.label),
                bottle.status.rawValue,

                CSVWriter.flag(bottle.isSample),
                CSVWriter.text(bottle.sampleFrom),
                CSVWriter.text(bottle.sampleSource?.rawValue),
                CSVWriter.flag(bottle.isInfinity),

                CSVWriter.decimal(bottle.volumeMl, places: 0),
                CSVWriter.decimal(bottle.abv),
                CSVWriter.decimal(bottle.abv.map { ABV(percent: $0).proof }),
                bottle.chillFiltered.map { $0 ? "yes" : "no" } ?? "",

                CSVWriter.text(bottle.batchNumber),
                CSVWriter.text(bottle.barrelNumber),
                CSVWriter.flag(bottle.isStorePick),
                CSVWriter.text(bottle.pickGroup),
                CSVWriter.text(bottle.pickStore),
                CSVWriter.text(bottle.warehouse),
                CSVWriter.text(bottle.rick),
                CSVWriter.text(bottle.floor),
                CSVWriter.text(bottle.recipeCode),
                CSVWriter.number(bottle.ageMonths),
                CSVWriter.decimal(bottle.entryProof),
                CSVWriter.number(bottle.charLevel),
                CSVWriter.text(bottle.finish),
                CSVWriter.number(bottle.bottleNumber),
                CSVWriter.number(bottle.bottlesInBatch),
                CSVWriter.text(bottle.topperLetter),
                CSVWriter.text(bottle.waxColor?.rawValue),
                CSVWriter.number(bottle.dripFraction.map { Int(($0 * 100).rounded()) }),
                CSVWriter.text(bottle.staveRecipe),
                CSVWriter.text(bottle.dsp),

                CSVWriter.number(bottle.distilledYear),
                CSVWriter.number(bottle.bottledYear),
                CSVWriter.number(bottle.vintageYear),

                CSVWriter.date(millis: bottle.purchaseDate),
                CSVWriter.date(millis: bottle.openedAt),
                CSVWriter.date(millis: bottle.finishedAt),
                CSVWriter.date(millis: summary.lastPouredAt.map {
                    Int64($0.timeIntervalSince1970 * 1000)
                }),
                CSVWriter.date(millis: bottle.lastVerifiedAt),

                CSVWriter.money(cents: bottle.purchasePriceCents),
                CSVWriter.text(bottle.purchaseStore),
                CSVWriter.text(bottle.storageLocation),
                CSVWriter.number(bottle.shelfNumber),

                CSVWriter.number(summary.status.remainingPours),
                CSVWriter.number(summary.status.totalPours),
                CSVWriter.decimal(summary.status.remainingMilliliters, places: 0),
                CSVWriter.money(cents: summary.costPerPourCents),

                CSVWriter.number(latest?.tasting.rating),
                CSVWriter.text(latest?.tasting.wouldRebuy?.rawValue),
                CSVWriter.number(latest?.tasting.finishSeconds),
                CSVWriter.text(latest?.tasting.liked),
                CSVWriter.text(latest?.tasting.disliked),
            ]
        }

        return CSVWriter.document(header: Self.header, rows: rows)
    }

    // MARK: - The rest of the record

    /// Every tasting, one row each, with the wheel picks per stage. The
    /// bottle export carries only the latest tasting; a person with three
    /// years of notes on one bottle needs all three.
    public static let tastingsHeader = [
        "date", "bottle", "product", "distillery", "rating", "blind",
        "would_rebuy", "worth_the_price", "heat", "finish_seconds",
        "source", "source_note", "liked", "disliked",
        "nose", "entry", "mid", "finish",
    ]

    public func tastingsCSV(
        resolveName: (Bottle) -> String,
        resolveIdentity: (String) -> ProductIdentity?,
        word: (String) -> String = { $0 }
    ) throws -> String {
        let bottles = BottleRepository(db)
        let names = Dictionary(
            try bottles.summaries(includeFinished: true).map { ($0.id, resolveName($0.bottle)) },
            uniquingKeysWith: { a, _ in a })
        let rows: [[String]] = try TastingRepository(db).allDetails().map { detail in
            let t = detail.tasting
            let identity = t.catalogProductId.flatMap(resolveIdentity)
            func stage(_ s: TastingStage) -> String {
                detail.descriptors(on: s).map(word).joined(separator: "; ")
            }
            return [
                CSVWriter.date(millis: t.tastedAt),
                CSVWriter.text(t.bottleId.flatMap { names[$0] }),
                CSVWriter.text(identity?.displayName),
                CSVWriter.text(identity?.distillery),
                CSVWriter.number(t.rating),
                CSVWriter.flag(t.blind),
                CSVWriter.text(t.wouldRebuy?.rawValue),
                t.worthThePrice.map { $0 ? "yes" : "no" } ?? "",
                CSVWriter.number(t.perceivedHeat),
                CSVWriter.number(t.finishSeconds),
                CSVWriter.text(t.source?.rawValue),
                CSVWriter.text(t.sourceNote),
                CSVWriter.text(t.liked),
                CSVWriter.text(t.disliked),
                stage(.nose), stage(.entry), stage(.mid), stage(.finish),
            ]
        }
        return CSVWriter.document(header: Self.tastingsHeader, rows: rows)
    }

    /// Every pour, one row each: the record the fill levels are derived
    /// from, and who a pour went to.
    public static let poursHeader = ["date", "bottle", "ml", "given_to", "into", "note"]

    public func poursCSV(resolveName: (Bottle) -> String) throws -> String {
        let bottles = BottleRepository(db)
        let names = Dictionary(
            try bottles.summaries(includeFinished: true).map { ($0.id, resolveName($0.bottle)) },
            uniquingKeysWith: { a, _ in a })
        let rows: [[String]] = try bottles.pours().map { pour in
            [
                CSVWriter.date(millis: pour.pouredAt),
                CSVWriter.text(names[pour.bottleId]),
                CSVWriter.decimal(pour.volumeMl, places: 1),
                CSVWriter.text(pour.givenTo),
                CSVWriter.text(pour.intoBottleId.flatMap { names[$0] }),
                CSVWriter.text(pour.note),
            ]
        }
        return CSVWriter.document(header: Self.poursHeader, rows: rows)
    }

    /// The hunt log, one row per sighting or lottery entry.
    public static let huntLogHeader = [
        "date", "kind", "product", "store", "region", "price", "count", "outcome", "bought", "note",
    ]

    public func huntLogCSV(
        resolveIdentity: (String) -> ProductIdentity?
    ) throws -> String {
        let rows: [[String]] = try SightingRepository(db).all().map { row in
            [
                CSVWriter.date(millis: row.seenAt),
                row.kind.rawValue,
                CSVWriter.text(row.catalogProductId.flatMap(resolveIdentity)?.displayName ?? row.customName),
                row.store,
                CSVWriter.text(row.region),
                CSVWriter.money(cents: row.cents),
                CSVWriter.number(row.count),
                CSVWriter.text(row.outcome?.rawValue),
                CSVWriter.flag(row.bottleId != nil),
                CSVWriter.text(row.note),
            ]
        }
        return CSVWriter.document(header: Self.huntLogHeader, rows: rows)
    }

    /// The passport, one row per visit.
    public static let visitsHeader = ["date", "distillery", "note"]

    public func visitsCSV() throws -> String {
        let rows: [[String]] = try VisitRepository(db).all().map { row in
            [CSVWriter.date(millis: row.visitedAt), row.distillery, CSVWriter.text(row.note)]
        }
        return CSVWriter.document(header: Self.visitsHeader, rows: rows)
    }

    /// Everything, as files: the bottles always, and the tastings, pours,
    /// hunt log and visits when there are any. Returns the URLs for a share sheet.
    public func writeAll(
        to directory: URL,
        resolveName: (Bottle) -> String,
        resolveIdentity: (String) -> ProductIdentity?,
        word: (String) -> String = { $0 }
    ) throws -> [URL] {
        let stamp = ISO8601DateFormatter().string(from: Date()).prefix(10)
        var urls = [try write(to: directory, resolveName: resolveName, resolveIdentity: resolveIdentity)]
        let extras: [(String, String)] = [
            ("tastings", try tastingsCSV(resolveName: resolveName, resolveIdentity: resolveIdentity, word: word)),
            ("pours", try poursCSV(resolveName: resolveName)),
            ("hunt-log", try huntLogCSV(resolveIdentity: resolveIdentity)),
            ("visits", try visitsCSV()),
        ]
        // A header alone is one line; anything recorded makes two. (CRLF is
        // one Character in Swift, so this is a string split, not a Character one.)
        for (name, contents) in extras where contents.components(separatedBy: "\r\n").filter({ !$0.isEmpty }).count > 1 {
            let url = directory.appendingPathComponent("pour-memo-\(name)-\(stamp).csv")
            try contents.write(to: url, atomically: true, encoding: .utf8)
            urls.append(url)
        }
        return urls
    }

    /// Writes the export to a file and returns its URL, for a share sheet.
    public func write(
        to directory: URL,
        resolveName: (Bottle) -> String,
        resolveIdentity: (String) -> ProductIdentity?
    ) throws -> URL {
        let contents = try csv(resolveName: resolveName, resolveIdentity: resolveIdentity)
        let stamp = ISO8601DateFormatter().string(from: Date()).prefix(10)
        let url = directory.appendingPathComponent("pour-memo-\(stamp).csv")
        try contents.write(to: url, atomically: true, encoding: .utf8)
        return url
    }

    /// What the user has, for a bottle chooser. Last-poured comes from the pour
    /// log rather than a stored column, like every other derived number here.
    public func pourCandidates(
        resolveName: (Bottle) -> String = { $0.customName ?? $0.id }
    ) throws -> [PickMyPour.Candidate] {
        try BottleRepository(db).summaries().map { summary in
            PickMyPour.Candidate(
                id: summary.id,
                name: resolveName(summary.bottle),
                lastPouredAt: summary.lastPouredAt,
                isOpen: summary.bottle.isOpen,
                remainingMilliliters: summary.status.remainingMilliliters)
        }
    }
}

File: Android/data/src/main/kotlin/com/talastack/liquorlog/data/CollectionExport.kt
package com.talastack.liquorlog.data

import com.talastack.liquorlog.engine.ABV
import com.talastack.liquorlog.engine.CSVWriter

/**
 * Everything you recorded, as CSV.
 *
 * **Export is never charged for, and never will be.** It costs almost
 * nothing and it is the single strongest trust signal in a category where
 * people have been burned by apps that held their data hostage. Now that
 * nothing in this app is charged for at all, the principle is simply the
 * shape of the feature: it is here, it is complete, and it runs with no
 * account.
 *
 * The columns are the iOS ones, in the iOS order, so a file exported on one
 * phone imports on the other. Deriving them separately per platform is how
 * two column sets that are meant to be identical drift apart.
 */
class CollectionExport(private val database: LiquorDatabase) {

    private val bottles = BottleRepository(database)
    private val tastings = TastingRepository(database)

    companion object {
        /** One row per bottle. Matches `CollectionExport.header` in Swift. */
        val bottleHeader: List<String> = listOf(
            "name", "distillery", "brand", "expression",
            "class", "production", "status",
            "sample", "sample_from", "infinity",
            "size_ml", "abv", "proof",
            "batch", "barrel", "store_pick", "pick_group", "pick_store",
            "warehouse", "rick", "floor", "recipe_code",
            "age_months", "bottle_number", "bottles_in_batch", "topper_letter", "dsp",
            "distilled_year", "bottled_year", "vintage_year",
            "purchased", "opened", "killed", "last_poured",
            "price", "bought_at", "storage_location", "shelf_number",
            "pours_remaining", "pours_total", "ml_remaining", "cost_per_pour",
            "rating",
        )

        val tastingHeader: List<String> = listOf(
            "date", "bottle", "rating", "blind",
            "would_rebuy", "worth_the_price", "heat", "finish_seconds",
            "source", "source_note", "liked", "disliked",
            "nose", "entry", "mid", "finish",
        )

        val pourHeader: List<String> = listOf("date", "bottle", "ml", "given_to", "note")
    }

    /**
     * Every bottle, finished ones included.
     *
     * An export that quietly dropped what you have already drunk would be an
     * export of part of the record, and the person would not find out until
     * they needed the part that was missing.
     */
    fun bottlesCsv(
        name: (BottleRepository.Summary) -> String = { it.name },
        distillery: (BottleRepository.Summary) -> String? = { null },
        kind: (BottleRepository.Summary) -> Pair<String?, String?> = { null to null },
    ): String {
        val rows = bottles.all().map { summary ->
            val b = summary.bottle
            val (classLabel, productionLabel) = kind(summary)
            listOf(
                name(summary),
                CSVWriter.text(distillery(summary)),
                CSVWriter.text(b.custom_name),
                "",
                CSVWriter.text(classLabel),
                CSVWriter.text(productionLabel),
                status(summary),
                CSVWriter.flag(summary.isSample),
                CSVWriter.text(b.sample_from),
                CSVWriter.flag(summary.isInfinity),
                CSVWriter.decimal(b.volume_ml, places = 0),
                CSVWriter.decimal(b.abv, places = 1),
                CSVWriter.decimal(b.abv?.let { ABV(it).proof }, places = 1),
                CSVWriter.text(b.batch_number),
                CSVWriter.text(b.barrel_number),
                CSVWriter.flag(summary.isStorePick),
                CSVWriter.text(b.pick_group),
                CSVWriter.text(b.pick_store),
                CSVWriter.text(b.warehouse),
                CSVWriter.text(b.rick),
                CSVWriter.text(b.floor),
                CSVWriter.text(b.recipe_code),
                CSVWriter.number(b.age_months?.toInt()),
                CSVWriter.number(b.bottle_number?.toInt()),
                CSVWriter.number(b.bottles_in_batch?.toInt()),
                CSVWriter.text(b.topper_letter),
                CSVWriter.text(b.dsp),
                CSVWriter.number(b.distilled_year?.toInt()),
                CSVWriter.number(b.bottled_year?.toInt()),
                CSVWriter.number(b.vintage_year?.toInt()),
                CSVWriter.date(b.purchase_date),
                CSVWriter.date(b.opened_at),
                CSVWriter.date(b.finished_at),
                CSVWriter.date(summary.lastPouredAt),
                CSVWriter.money(b.purchase_price_cents?.toInt()),
                CSVWriter.text(b.purchase_store),
                CSVWriter.text(b.storage_location),
                CSVWriter.number(b.shelf_number?.toInt()),
                CSVWriter.number(summary.status.remainingPours),
                CSVWriter.number(summary.status.totalPours),
                CSVWriter.decimal(summary.status.remainingMilliliters, places = 0),
                CSVWriter.money(summary.costPerPourCents),
                CSVWriter.number(summary.latestRating),
            )
        }
        return CSVWriter.document(bottleHeader, rows)
    }

    /** Every tasting, with its descriptors spelled out per stage. */
    fun tastingsCsv(
        name: (String) -> String = { it },
        label: (String) -> String = { it },
    ): String {
        val rows = tastings.allDetails().map { detail ->
            val t = detail.tasting
            fun stage(s: TastingRepository.Stage) =
                detail.descriptors(s).joinToString("; ") { label(it) }
            listOf(
                CSVWriter.date(detail.tastedAt),
                CSVWriter.text(detail.bottleId?.let { name(it) }),
                CSVWriter.number(detail.rating),
                CSVWriter.flag(detail.isBlind),
                CSVWriter.text(detail.rebuy?.storageKey),
                // Null and "no" are different answers. `flag` cannot say the
                // first, so a nullable column is written by hand.
                t.worth_the_price?.let { if (it != 0L) "yes" else "no" } ?: "",
                CSVWriter.number(t.perceived_heat?.toInt()),
                CSVWriter.number(t.finish_seconds?.toInt()),
                CSVWriter.text(t.source),
                CSVWriter.text(t.source_note),
                CSVWriter.text(t.liked),
                CSVWriter.text(t.disliked),
                stage(TastingRepository.Stage.NOSE),
                stage(TastingRepository.Stage.ENTRY),
                stage(TastingRepository.Stage.MID),
                stage(TastingRepository.Stage.FINISH),
            )
        }
        return CSVWriter.document(tastingHeader, rows)
    }

    /** Every pour, which is where the fill levels come from. */
    fun poursCsv(name: (String) -> String = { it }): String {
        val rows = bottles.all().flatMap { summary ->
            bottles.poursFor(summary.id).map { pour ->
                listOf(
                    CSVWriter.date(pour.poured_at),
                    name(summary.id),
                    CSVWriter.decimal(pour.volume_ml, places = 0),
                    CSVWriter.text(pour.given_to),
                    CSVWriter.text(pour.note),
                )
            }
        }
        return CSVWriter.document(pourHeader, rows)
    }

    /**
     * A bottle's state as one word.
     *
     * Three states, not two: sealed, open and finished are what the collection
     * filter uses, and an export that collapsed them would not round-trip.
     */
    private fun status(summary: BottleRepository.Summary): String = when {
        summary.isFinished -> "finished"
        summary.isOpen -> "open"
        else -> "sealed"
    }
}
